internal/target/service_list_more.go

```diff
@@ -0,0 +1,111 @@
+
+import (
+	"context"
+	"time"
+
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/refreshtoken"
+	rt "github.com/hashicorp/boundary/internal/refreshtoken"
+)
+
+func ListMore(
+	ctx context.Context,
+	tok *rt.RefreshToken,
+	repo *Repository,
+	grantsHasher rt.GrantsHasher,
+	pageSize int,
+	filterItemFn func(Target) (bool, error),
+) (*ListResponse, error) {
+	const op = "target.ListMore"
+
+	deletedIds, transactionTimestamp, err := repo.ListDeletedIds(ctx, tok.CreatedTime)
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op)
+	}
+
+	limit := pageSize + 1
+	opts := []Option{
+		WithLimit(limit),
+		WithStartPageAfterItem(tok.LastItemId, tok.LastItemUpdatedTime),
+	}
+
+	// pagination magic
+	targets := make([]Target, 0, pageSize+1)
+dbLoop:
+	for {
+		// Request another page from the DB until we fill the final items
+		page, err := repo.ListTargets(ctx, opts...)
+		if err != nil {
+			return nil, errors.Wrap(ctx, err, op)
+		}
+		for _, item := range page {
+			ok, err := filterItemFn(item)
+			if err != nil {
+				return nil, errors.Wrap(ctx, err, op)
+			}
+			if ok {
+				targets = append(targets, item)
+				// If we filled the items after filtering,
+				// we're done.
+				if len(targets) == cap(targets) {
+					break dbLoop
+				}
+			}
+		}
+		// If the current page was shorter than the limit, stop iterating
+		if len(page) < limit {
+			break dbLoop
+		}
+
+		opts = []Option{
+			WithLimit(limit),
+			WithStartPageAfterItem(page[len(page)-1].GetPublicId(), page[len(page)-1].GetUpdateTime().AsTime()),
+		}
+	}
+	// If we couldn't fill the items, it was a complete listing.
+	completeListing := len(targets) < cap(targets)
+	if !completeListing {
+		// Items is of size pageSize+1, so
+		// truncate if it was filled.
+		targets = targets[:pageSize]
+	}
+
+	totalItems, err := repo.EstimatedCount(ctx)
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op)
+	}
+
+	grantsHash, err := grantsHasher.GrantsHash(ctx)
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op)
+	}
+
+	resp := &ListResponse{
+		Items:               targets,
+		DeletedIds:          deletedIds,
+		EstimatedTotalItems: totalItems,
+		CompleteListing:     completeListing,
+		RefreshToken: &refreshtoken.RefreshToken{
+			// Use the timestamp of the deleted IDs transaction with a
+			// buffer to account for overlapping transactions. It is okay
+			// to return a deleted ID more than once. The buffer corresponds
+			// to Postgres' default transaction timeout.
+			CreatedTime:         transactionTimestamp.Add(-30 * time.Second),
+			ResourceType:        refreshtoken.ResourceTypeTarget,
+			PermissionsHash:     grantsHash,
+			LastItemId:          tok.LastItemId,
+			LastItemUpdatedTime: tok.LastItemUpdatedTime,
+		},
+	}
+
+	if len(targets) > 0 {
+		resp.RefreshToken.LastItemId = targets[len(targets)-1].GetPublicId()
+		resp.RefreshToken.LastItemUpdatedTime = targets[len(targets)-1].GetUpdateTime().AsTime()
+	}
```

Comment:
I think this could be moved into a "Refresh" method on a RefreshToken. I think the only thing that should change between the refresh token that was passed in is the LastItemId and LastItemUpdatedTime. So similarly to the `New` function that was mentioned, you could have something like:

```go
func (r *RefreshToken) Refresh(res boundary.Resource) error {
  // validation like that resource type matches,
  // and that update time is newer than current update time

  r.LastItemId = res.GetPublicId()
  r.LastItemUpdatedTime = res.GetUpdateTime()
}
```

Reply:
I view refresh tokens as immutable, even though they are sometimes based on values from a previous refresh token, so I don't think this is the right function. Conceptually I think the grants hash should come from the auth result every time, not the previous token. Create time should come from the database, and the resource type is specific to the resource. Let me think about how to change this though.

Reply:
Done, thanks!